Allow to estimate free space when using object storage

tcitworld opened the pull request:
With ObjectStorage there's no way to determine the free space for a bucket with a limited storage quota, so it's hard to detect when you have a space issue (there's nothing like `disk_free_space`).

It would be better if the object storage provider would provide us this information, but at least for S3 it's hard or impossible to get this information easily, so this can be an appropriate workaround in the meanwhile.

~~Using an AppConfig instead of a SystemConfig might be more flexible for admins to change the value, but it might also introduce a database overhead.~~

When using the `Local` storage regular `disk_free_space` is used instead.

Still need to check if and how this affects external storages.

CarlSchwan commented:
I prefer using an app config, the overhead is minimal for such simple configs and the value of this config doesn't need to be know before loading the database

tcitworld committed:
Allow to mesure free space in object storages, based on used size and quota

If the object storage backend doesn't return quota or used bytes
information, use a config override for the quota and count the size used
in Nextcloud instead.

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

blizzz commented:
doesn't it overlap with the `default_quota`?

tcitworld commented:
`default_quota` is for users, this is for the whole instance. It should still work anyway.

PVince81 commented:
sounds like a good idea

would still like to have another opinion in case of side effects @icewind1991 @juliushaertl

skjnldsv reviewed (dismissed):
I would prefer a proper config flag. This is not discoverable as n app config :)
Or we add that in the settings UI somewhere

tcitworld commented:
I'm fine with anything. @CarlSchwan preferred an app config. :shrug:

skjnldsv commented:
@CarlSchwan what's the reasoning behind? :)
Maybe I missed some background behind that PR :thinking:

juliushaertl commented on lib/private/Files/Storage/Common.php:
```diff
@@ -474,12 +475,20 @@ public function test() {
 	 */
 	public function free_space($path) {
+		/** @var IConfig $config */
+		$config = \OC::$server->get(IConfig::class);
```
```suggestion
		$config = \OCP\Server::get(IConfig::class);
```

juliushaertl commented on lib/private/legacy/OC_Helper.php:
```diff
@@ -517,6 +518,15 @@ public static function getStorageInfo($path, $rootInfo = null, $includeMountPoin
+
+		// If the admin overrides the global instance quota we default to the value set
+		/** @var IConfig $config */
+		$config = \OC::$server->get(IConfig::class);
```
```suggestion
		$config = \OCP\Server::get(IConfig::class);
```

juliushaertl commented:
Looks good, but for external storage I think this value should not apply. 

Might be better to implement this part of free_space in the ObjectStoreStorage class and overload it in AmazonS3::free_space so that external S3 storage still returns unknown free space.

tcitworld commented:
I changed the scope of the PR. The aim is to obtain quota information and used size from the object storage. The `OCP\Files\ObjectStore\IObjectStore` interface now has two extra methods, `bytesUsed` and `bytesQuota`.

As far as I could find, only OpenStack Swift exposes used size and quota information directly through their API, so that's the only implementation that doesn't return a default value.

When we don't get given quota information from the object storage backend, we check if it's overwritten by the new `quota_instance_global` config key.

If there's a quota (`> 0`) in place, we can start using the free space given from the object storage backend and simply substract the used bytes to the quota to get the free space.

When the object storage backend doesn't return used size, we sum the size of every object storage "storage" registered and given through the file cache root. This should give an approximative value that shouldn't be far from the truth, provided:
* the bucket isn't used for anything else (which Nextcloud already assumes)
* file versioning isn't handled by the object storage backend like with `files_versions_s3`)

icewind1991 commented on lib/private/Files/ObjectStore/ObjectStoreStorage.php:
```diff
@@ -627,4 +631,68 @@ private function copyFile(ICacheEntry $sourceEntry, string $to) {
+		if ($quota < 0) {
+			/** @var IConfig $config */
+			$config = Server::get(IConfig::class);
+			$quota = (int) $config->getAppValue('core', 'quota_instance_global', '0');
```
Probably makes sense to move this option under the rest of the object store options, since it is only applicable for object stores

tcitworld replied:
You mean the `objectstore` config.php entries? Makes sense indeed, will do.

skjnldsv commented:
Closing this pull request due to lack of recent activity and updates. We appreciate your contribution and encourage you to reopen or provide further updates if necessary. :relaxed: 
Our aim is to keep the project moving forward with active collaboration. Thank you for your understanding and continued support! :pray:

tcitworld commented:
https://github.com/nextcloud/server/pull/54797 seems to handle it properly
